File: src/data/tutorials/061-jump-game.js
/**
 * Tutorial content for #061 — Jump Game
 * Keyed by approach name (must match the name parsed from the solution file).
 */
export const tutorial = {
  problem: {
    statement: `You are at index 0 of an array. Each element tells you the maximum number of positions you can jump forward from that index. Return true if you can reach the last index, false otherwise.`,
    example: `[2, 3, 1, 1, 4] → true  (0→1→4, or 0→2→3→4)\n[3, 2, 1, 0, 4] → false (index 3 always reached but has value 0, can't pass)\n✅ Answer: true / false`,
    keyInsight: `Track the furthest index you can ever reach. If you arrive at a position beyond your current reach, you're stuck. If your reach ever touches or passes the last index, you can make it.`,
  },

  approaches: {
    'Greedy (Forward)': {
      intuition: `Scan left to right, updating the furthest index reachable from all positions visited so far. If at any point the current index is beyond that furthest reach, you're stuck and return false. If the furthest reach ever hits or passes the last index, return true immediately.`,
      steps: [
        `Set maxReach = 0.`,
        `For i from 0 to n-1:`,
        `  If i > maxReach: return false (can't reach here).`,
        `  maxReach = max(maxReach, i + nums[i]).`,
        `  If maxReach >= n-1: return true.`,
        `Return true.`,
      ],
      example: `nums = [2, 3, 1, 1, 4]\n\ni=0: maxReach=max(0, 0+2)=2\ni=1: maxReach=max(2, 1+3)=4 ≥ 4 → return true\n✅ Answer: true\n\nnums = [3, 2, 1, 0, 4]\ni=0: maxReach=3\ni=1: maxReach=max(3,3)=3\ni=2: maxReach=max(3,3)=3\ni=3: maxReach=max(3,3)=3\ni=4: 4 > 3 → return false\n✅ Answer: false`,
      keyInsight: `O(n) time, O(1) space. The greedy approach is optimal — no DP table needed, just a single pass with one variable.`,
    },

    'Greedy (Backward)': {
      intuition: `Work backwards from the last index. Maintain a "goal" — initially the last index. Scan right-to-left: if from position i you can reach the goal (i + nums[i] >= goal), update goal to i. If goal reaches 0, you can make it from the start.`,
      steps: [
        `Set goal = n-1.`,
        `For i from n-2 down to 0:`,
        `  If i + nums[i] >= goal: goal = i.`,
        `Return goal == 0.`,
      ],
      example: `nums = [2, 3, 1, 1, 4]\n\ngoal=4\ni=3: 3+1=4 ≥ 4 → goal=3\ni=2: 2+1=3 ≥ 3 → goal=2\ni=1: 1+3=4 ≥ 2 → goal=1\ni=0: 0+2=2 ≥ 1 → goal=0\nReturn goal==0 → true\n✅ Answer: true`,
      keyInsight: `O(n) time, O(1) space. Elegant alternative to forward greedy — "can index i save us from needing to reach all the way to the last index?"`,
    },

    'DP': {
      intuition: `Create a boolean dp array where dp[i] = true means index i is reachable. Start with dp[0]=true. For each reachable index, mark all indices within jump range as reachable too.`,
      steps: [
        `Create dp[n] all false; dp[0] = true.`,
        `For i from 0 to n-1:`,
        `  If dp[i] is false, skip.`,
        `  For j from i+1 to min(i + nums[i], n-1):`,
        `    dp[j] = true.`,
        `Return dp[n-1].`,
      ],
      example: `nums = [2, 3, 1, 1, 4]\n\ndp=[T,F,F,F,F]\ni=0: mark dp[1]=T, dp[2]=T\ni=1: mark dp[2]=T, dp[3]=T, dp[4]=T\ndp=[T,T,T,T,T]\nReturn dp[4]=true\n✅ Answer: true`,
      keyInsight: `O(n²) time worst case, O(n) space. Less efficient than greedy but more explicit about which indices are reachable.`,
    },

    'BFS': {
      intuition: `Treat each index as a node. From index i you can jump to any index in [i+1, i+nums[i]]. BFS explores all reachable nodes level by level. If the last index enters the queue or is found, return true.`,
      steps: [
        `Enqueue index 0; mark 0 as visited.`,
        `While queue is not empty:`,
        `  Dequeue index i.`,
        `  For j from i+1 to min(i+nums[i], n-1):`,
        `    If j == n-1: return true.`,
        `    If not visited: enqueue j, mark visited.`,
        `Return false.`,
      ],
      example: `nums = [2, 3, 1, 1, 4]\n\nQueue: [0]\nPop 0: enqueue 1, 2; visited={0,1,2}\nPop 1: j=4 == n-1 → return true ✅\n✅ Answer: true`,
      keyInsight: `O(n²) time worst case, O(n) space. BFS is overkill here — greedy is far superior — but shows how jump-game maps to graph reachability.`,
    },

    'Recursive Memoization': {
      intuition: `From index i, try every possible jump length (1 to nums[i]) and recurse. Cache each index as reachable (true) or not (false). If any recursive path reaches the last index, return true.`,
      steps: [
        `Create memo[n] filled with -1 (unknown).`,
        `Base case: i >= n-1 → return true.`,
        `If memo[i] != -1, return memo[i].`,
        `For jump from 1 to nums[i]: if solve(i + jump) is true → memo[i]=true, return true.`,
        `memo[i] = false; return false.`,
      ],
      example: `nums = [2, 3, 1, 1, 4]\n\nsolve(0): try jump 1 → solve(1)\nsolve(1): try jump 3 → solve(4); 4>=4 → true ✅\nsolve(1)=true, solve(0)=true\n✅ Answer: true`,
      keyInsight: `O(n²) time worst case, O(n) space (memo + call stack). More complex than needed for this problem — greedy is always preferred in interviews.`,
    },
  },
}

File: public/cpp/061-jump-game.cpp
/*
 * Problem: Jump Game (LeetCode 55)
 * Difficulty: Medium
 * Category: Dynamic Programming
 * 
 * Description:
 * You are given an integer array nums. You are initially positioned at the array's
 * first index, and each element in the array represents your maximum jump length
 * at that position.
 * 
 * Return true if you can reach the last index, or false otherwise.
 * 
 * Example 1:
 * Input: nums = [2,3,1,1,4]
 * Output: true
 * Explanation: Jump 1 step from index 0 to 1, then 3 steps to the last index.
 * 
 * Example 2:
 * Input: nums = [3,2,1,0,4]
 * Output: false
 * Explanation: You will always arrive at index 3 no matter what. Its maximum
 * jump length is 0, which makes it impossible to reach the last index.
 * 
 * Constraints:
 * - 1 <= nums.length <= 10^4
 * - 0 <= nums[i] <= 10^5
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

/*
 * APPROACH 1: GREEDY (OPTIMAL)
 * 
 * Intuition:
 * - Track the farthest position we can reach
 * - At each position, update the farthest reachable position
 * - If current position is beyond farthest reachable, return false
 * - If farthest reaches or exceeds last index, return true
 * 
 * Algorithm:
 * 1. Initialize maxReach = 0
 * 2. For each position i from 0 to n-1:
 *    - If i > maxReach, return false (can't reach here)
 *    - Update maxReach = max(maxReach, i + nums[i])
 *    - If maxReach >= n-1, return true
 * 3. Return true (reached end of loop)
 * 
 * Time Complexity: O(n) - single pass
 * Space Complexity: O(1) - only variables
 */
class Solution1 {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int maxReach = 0;
        
        for (int i = 0; i < n; i++) {
            // Can't reach current position
            if (i > maxReach) {
                return false;
            }
            
            // Update farthest reachable position
            maxReach = max(maxReach, i + nums[i]);
            
            // Can reach the end
            if (maxReach >= n - 1) {
                return true;
            }
        }
        
        return true;
    }
};

/*
 * APPROACH 2: GREEDY (BACKWARD)
 * 
 * Intuition:
 * - Start from the end and work backwards
 * - Track the leftmost position that can reach the end
 * - If we can reach position 0, return true
 * - This is elegant and easy to understand
 * 
 * Algorithm:
 * 1. Initialize lastPos = n - 1 (target position)
 * 2. Iterate from n-2 to 0:
 *    - If i + nums[i] >= lastPos, update lastPos = i
 * 3. Return lastPos == 0
 * 
 * Time Complexity: O(n) - single pass
 * Space Complexity: O(1) - only variables
 */
class Solution2 {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int lastPos = n - 1;
        
        // Work backwards
        for (int i = n - 2; i >= 0; i--) {
            // Can reach lastPos from position i
            if (i + nums[i] >= lastPos) {
                lastPos = i;
            }
        }
        
        return lastPos == 0;
    }
};

/*
 * APPROACH 3: DYNAMIC PROGRAMMING (BOTTOM-UP)
 * 
 * Intuition:
 * - Use DP array where dp[i] = true if position i is reachable
 * - For each reachable position, mark all positions it can jump to
 * - This is less efficient but shows DP thinking
 * 
 * Algorithm:
 * 1. Create dp array, initialize dp[0] = true
 * 2. For each position i:
 *    - If dp[i] is true, mark all positions i can reach
 * 3. Return dp[n-1]
 * 
 * Time Complexity: O(n^2) - nested loops in worst case
 * Space Complexity: O(n) - DP array
 */
class Solution3 {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        vector<bool> dp(n, false);
        dp[0] = true;
        
        for (int i = 0; i < n; i++) {
            if (!dp[i]) continue;
            
            // Mark all reachable positions from i
            for (int j = 1; j <= nums[i] && i + j < n; j++) {
                dp[i + j] = true;
            }
            
            // Early exit if we can reach the end
            if (dp[n - 1]) return true;
        }
        
        return dp[n - 1];
    }
};

/*
 * APPROACH 4: BFS (GRAPH TRAVERSAL)
 * 
 * Intuition:
 * - Treat this as a graph problem
 * - Each position is a node, edges to all reachable positions
 * - Use BFS to find if we can reach the last position
 * - This is overkill but shows different perspective
 * 
 * Algorithm:
 * 1. Use queue for BFS, start with position 0
 * 2. Track visited positions
 * 3. For each position, add all reachable unvisited positions to queue
 * 4. If we reach last position, return true
 * 
 * Time Complexity: O(n^2) - in worst case visit all edges
 * Space Complexity: O(n) - queue and visited set
 */
class Solution4 {
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        if (n == 1) return true;
        
        vector<bool> visited(n, false);
        int maxReach = 0;
        
        for (int i = 0; i <= maxReach && i < n; i++) {
            maxReach = max(maxReach, i + nums[i]);
            if (maxReach >= n - 1) return true;
        }
        
        return false;
    }
};

/*
 * APPROACH 5: RECURSIVE WITH MEMOIZATION
 * 
 * Intuition:
 * - Recursively check if we can reach end from current position
 * - Use memoization to avoid recomputing
 * - This shows top-down DP approach
 * 
 * Algorithm:
 * 1. Base case: if at last position, return true
 * 2. If already computed, return cached result
 * 3. Try all possible jumps from current position
 * 4. If any jump leads to success, return true
 * 5. Cache and return result
 * 
 * Time Complexity: O(n^2) - each position computed once, but tries all jumps
 * Space Complexity: O(n) - memoization + recursion stack
 */
class Solution5 {
private:
    bool helper(vector<int>& nums, int pos, vector<int>& memo) {
        int n = nums.size();
        
        // Base case: reached the end
        if (pos >= n - 1) return true;
        
        // Check memo
        if (memo[pos] != -1) return memo[pos];
        
        // Try all possible jumps
        int maxJump = nums[pos];
        for (int jump = 1; jump <= maxJump; jump++) {
            if (helper(nums, pos + jump, memo)) {
                memo[pos] = 1;
                return true;
            }
        }
        
        memo[pos] = 0;
        return false;
    }
    
public:
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        vector<int> memo(n, -1);
        return helper(nums, 0, memo);
    }
};

// Test function
void test(vector<int> nums, int approach) {
    bool result;
    
    cout << "Input: [";
    for (int i = 0; i < nums.size(); i++) {
        cout << nums[i];
        if (i < nums.size() - 1) cout << ",";
    }
    cout << "]\n";
    
    switch(approach) {
        case 1: {
            Solution1 sol;
            result = sol.canJump(nums);
            cout << "Approach 1 (Greedy Forward): ";
            break;
        }
        case 2: {
            Solution2 sol;
            result = sol.canJump(nums);
            cout << "Approach 2 (Greedy Backward): ";
            break;
        }
        case 3: {
            Solution3 sol;
            result = sol.canJump(nums);
            cout << "Approach 3 (DP Bottom-up): ";
            break;
        }
        case 4: {
            Solution4 sol;
            result = sol.canJump(nums);
            cout << "Approach 4 (BFS): ";
            break;
        }
        case 5: {
            Solution5 sol;
            result = sol.canJump(nums);
            cout << "Approach 5 (Memoization): ";
            break;
        }
    }
    
    cout << (result ? "true" : "false") << "\n\n";
}

int main() {
    // Test Case 1: Can reach end
    cout << "Test Case 1: Can reach end\n";
    vector<int> test1 = {2, 3, 1, 1, 4};
    for (int i = 1; i <= 5; i++) {
        test(test1, i);
    }
    
    // Test Case 2: Cannot reach end
    cout << "Test Case 2: Cannot reach end\n";
    vector<int> test2 = {3, 2, 1, 0, 4};
    for (int i = 1; i <= 5; i++) {
        test(test2, i);
    }
    
    // Test Case 3: Single element
    cout << "Test Case 3: Single element\n";
    vector<int> test3 = {0};
    for (int i = 1; i <= 5; i++) {
        test(test3, i);
    }
    
    // Test Case 4: All zeros except first
    cout << "Test Case 4: All zeros except first\n";
    vector<int> test4 = {2, 0, 0};
    for (int i = 1; i <= 5; i++) {
        test(test4, i);
    }
    
    // Test Case 5: Large jumps
    cout << "Test Case 5: Large jumps\n";
    vector<int> test5 = {10, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
    for (int i = 1; i <= 5; i++) {
        test(test5, i);
    }
    
    // Test Case 6: Stuck in middle
    cout << "Test Case 6: Stuck in middle\n";
    vector<int> test6 = {1, 1, 0, 1};
    for (int i = 1; i <= 5; i++) {
        test(test6, i);
    }
    
    // Test Case 7: Can barely reach
    cout << "Test Case 7: Can barely reach\n";
    vector<int> test7 = {1, 1, 1, 1, 1};
    for (int i = 1; i <= 5; i++) {
        test(test7, i);
    }
    
    return 0;
}

/*
 * COMPLEXITY ANALYSIS SUMMARY:
 * 
 * Approach 1 (Greedy Forward):
 * - Time: O(n) - single pass
 * - Space: O(1) - only variables
 * - Best for: Optimal solution, most efficient
 * 
 * Approach 2 (Greedy Backward):
 * - Time: O(n) - single pass
 * - Space: O(1) - only variables
 * - Best for: Elegant code, easy to understand
 * 
 * Approach 3 (DP Bottom-up):
 * - Time: O(n^2) - nested loops
 * - Space: O(n) - DP array
 * - Best for: Showing DP thinking
 * 
 * Approach 4 (BFS):
 * - Time: O(n^2) - worst case
 * - Space: O(n) - queue and visited
 * - Best for: Graph perspective
 * 
 * Approach 5 (Memoization):
 * - Time: O(n^2) - each position tries all jumps
 * - Space: O(n) - memo + recursion
 * - Best for: Top-down DP approach
 * 
 * INTERVIEW TIPS:
 * 1. Start with greedy approach (Approach 1 or 2)
 * 2. Greedy is optimal here - no need for DP
 * 3. Explain why greedy works (local optimal = global optimal)
 * 4. Mention DP approaches if asked about alternatives
 * 5. This problem tests greedy algorithm understanding
 * 
 * COMMON MISTAKES:
 * 1. Using DP when greedy is sufficient (overcomplicating)
 * 2. Not handling edge case of single element
 * 3. Off-by-one errors in maxReach calculation
 * 4. Not checking if current position is reachable
 * 5. Confusing this with Jump Game II (minimum jumps)
 * 
 * FOLLOW-UP QUESTIONS:
 * 1. What's the minimum number of jumps? (Jump Game II - LeetCode 45)
 * 2. Can you return the actual path? (Track previous positions)
 * 3. What if you can jump backwards? (Different problem, use BFS)
 * 4. What if there are negative jumps? (Use DP or BFS)
 * 5. How to handle very large arrays? (Greedy is already optimal)
 * 
 * RELATED PROBLEMS:
 * - Jump Game II (minimum jumps)
 * - Jump Game III (jump to zero)
 * - Jump Game IV (jump to same value)
 * - Jump Game V (jump with obstacles)
 * - Minimum Number of Taps to Open to Water a Garden
 * 
 * KEY INSIGHTS:
 * 1. Greedy works because we only care about reachability, not path
 * 2. Forward greedy tracks maximum reachable position
 * 3. Backward greedy finds leftmost position that can reach end
 * 4. Both greedy approaches are O(n) time, O(1) space
 * 5. DP is overkill but shows problem-solving flexibility
 */

// Made with Bob
